handout/testing/unit_testing_redux.md

```diff
@@ -0,0 +1,229 @@
+import {INCREMENT_COUNTER, DECREMENT_COUNTER} from './counter';
+import * as CounterActions from './counter';
+
+describe('counter action creators', () => {                     
+  ...
+
+  it('incrementIfOdd should dispatch INCREMENT_COUNTER if counter is odd', (done) => {
+    const expectedAction = { type: INCREMENT_COUNTER };
+
+    const store = mockStore({                          
+      getState: () => {                                
+        return {                                       
+          counter: 1                                   
+        }                                              
+      },                                               
+      dispatch: (action) => {                          
+        chai.expect(action)                            
+          .to.deep.equal(expectedAction);              
+
+        done();                                        
+      }                                                
+    });                                                
+
+    store.dispatch(CounterActions.incrementIfOdd());   
+  });
+});
+```
+
+We create a mock store with the state's `counter` set to an odd number and expect `incrementIfOdd` to indeed `INCREMENT_COUNTER`. The mock `dispatch` method checks for the action we expect to be created and dispatched.
+
+The tricky part is that `CounterActions.incrementIfOdd` is a thunk, so the `store.dispatch` above is actually handled by the `thunk` middleware and *not* the `dispatch` we pass into our `mockStore` helper, which is expecting a generic JSON object.
+
+With a few adjustments we can also test thunks that dispatch more than one action.
+
+```
+it('incrementIfOdd should dispatch INCREMENT_COUNTER if counter is odd', (done) => {
+  const expectedActions = [
+   { type: INCREMENT_COUNTER },
+   { type: DECREMENT_COUNTER },
+   { type: INCREMENT_COUNTER },
+   { type: DECREMENT_COUNTER },
+   { type: INCREMENT_COUNTER },
+   { type: DECREMENT_COUNTER }
+  ];
+
+  const store = mockStore({                          
+    getState: () => {                                
+      return {                                       
+        counter: 1                                   
+      }                                              
+    },                                               
+    dispatch: (action) => {                          
+      const expectedAction = expectedActions.shift();
+
+      chai.expect(action)                            
+        .to.deep.equal(expectedAction);
+
+      if (expectedAction.length === 0) {
+        done();                                        
+      };
+    }                                                
+  });                                                
+
+  // some thunk that dispatches our expectedActions
+  store.dispatch(CounterActions.zigZagZigala());   
+});
+```
+
+There's more than one flavor of asynchronous action creating in redux, so you might have to cook something that fits your unit's needs.  
```

Comment:
This sentence could be more formal.  Something like:

These asynchronous examples are based on the `thunk` pattern.  Redux is very flexible and there are many other ways to implement asynchronous actions, `thunk`s are merely the most common.  Other asynchronous implementations may require a different approach to unit testing.